compiler/hie-core/src/Development/IDE/Core/Debouncer.hs

```diff
@@ -0,0 +1,64 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module Development.IDE.Core.Debouncer
+    ( Debouncer
+    , newDebouncer
+    , withDebouncer
+    , registerEvent
+    , killEvents
+    , waitForEvents
+    ) where
+
+import Control.Concurrent.Extra
+import Control.Concurrent.Async
+import Control.Exception
+import Control.Monad.Extra
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import System.Time.Extra
+
+-- | A debouncer can be used to avoid triggering many events
+-- (e.g. diagnostics) for the same key (e.g. the same file)
+-- within a short timeframe. This is accomplished
+-- by delaying each event for a given time. If another event
+-- is registered for the same key within that timeframe,
+-- only the new event will fire.
+newtype Debouncer k = Debouncer (Var (Map k (Async ())))
+
+-- | Create a new empty debouncer.
+newDebouncer :: IO (Debouncer k)
+newDebouncer = do
+    m <- newVar Map.empty
+    pure $ Debouncer m
+
+-- | Operate on a bouncer and wait for all registered
+-- events to fire.
+withDebouncer :: (Debouncer k -> IO a) -> IO a
+withDebouncer = bracket newDebouncer waitForEvents
+
+-- | Register an event that will fire after the given delay if no other event
+-- for the same key gets registered until then.
+--
+-- If there is a pending event for the same key, the pending event will be killed.
+-- Events are run unmasked so it is up to the user of `registerEvent`
+-- to mask if required.
+registerEvent :: Ord k => Debouncer k -> Seconds -> k -> IO () -> IO ()
+registerEvent (Debouncer d) delay k fire = modifyVar_ d $ \m -> mask_ $ do
+    whenJust (Map.lookup k m) cancel
+    a <- asyncWithUnmask $ \unmask ->
+        (unmask $ sleep delay >> fire) `finally`
+        modifyVar_ d (pure . Map.delete k)
+    pure $ Map.insert k a m
+
+-- | Kill all registered events and clear the map.
+killEvents :: Debouncer k -> IO ()
+killEvents (Debouncer d) = modifyVar_ d $ \m -> do
+    mapM_ cancel $ Map.elems m
+    pure Map.empty
+
+-- | Wait for all registered events to fire.
+waitForEvents :: Debouncer k -> IO ()
+waitForEvents (Debouncer d) =
+    withVar d $ \m -> mapM_ wait $ Map.elems m
```

Comment:
This is OK, but a little sad that you always wait the 0.1 even though you know you aren't getting any new things coming in. Not easy to make it better without making it more complex though. Best I can suggest is having a `(Async (), IO ())` where the `IO ()` is a `once` that both try running. However, that's a lot of complexity for something not particularly valuable.

Reply:
I’ve just removed `waitForEvents` since we don’t actually use it.

Reply:
Oh, even better!